docs/rfcs/20220209_leader_epoch_support_rfc.md

```diff
@@ -0,0 +1,114 @@
+and update metadata against different broker. 
+
+## How (short plan)
+
+We will expose raft group terms, that are already gathered in redpanda metadata 
+infrastructure as leader epochs. Additionally we will implement 
+`OffsetForLeaderEpoch` Kafka API to provide full compatibility. 
+
+
+# Motivation
+## Why are we doing this?
+
+Adding leader epoch support will bring allow us to better handle stale metadata 
+and unclean truncations.
+
+### Note on unclean truncations in redpanda
+
+Since Redpanda replication is based on Raft we do not allow for unclean leader 
+election that may lead to high watermark being decreased. The only situation in 
+which high watermark may be decreased is simultaneous failure of majority of 
+nodes handling relaxed consistency writes such as `ACKS=1` and `ACKS=0`. 
+
+## What use cases does it support?
+
+We will add support for leader epoch in each request that requires it and is 
+currently handled by Redpanda.
+
+## What is the expected outcome?
+
+Validation of leader epoch that is consistent with Kafka.
+
+# Guide-level explanation
+## How do we teach this?
+
+This is a standard feature of Kafka protocol, we may use existing, related KIPS.
+
+- KIP-101 Alter Replication Protocol to use Leader Epoch rather than High Watermark for Truncation
+- KIP-232 Detect outdated metadata using leaderEpoch and partitionEpoch
+- KIP-320 Allow fetchers to detect and handle log truncation
+- KIP-359 Verify leader epoch in produce requests
+- KIP-359 Verify leader epoch in produce requests
+
+
+# Reference-level explanation
+
+## Detailed design - What needs to change to get there
+
+### Leader epoch metadata
+
+We need to store last stable leader term in `cluster::partition_leaders_table`.
+The term will directly be mapped to leader epoch returned in `MetadataResponse`. 
+
+We are going to introduce a type representing current leader and the term. 
+
+```c++
+    struct leader_term {
+        std::optional<model::node_id> leader;
+        model::term_id term;
+    };
+
+```
+
+### Offset for leader epoch API handler
+
+The `OffsetForLeaderEpoch` API returns last offset in request leader epoch plus 
+one. That is the first offset of next leader epoch or in other words an offset 
+on which requested leader epoch ends.
+
+The `OffsetForLeaderEpoch` is handled only by the requested partition leader. 
+In order to implement that feature we need to add an API that will allow us to 
+query for last offset in given term. Then the offset will have to be translated 
+with an offset translator to obtain valid Kafka offset.
+
+For local partitions this is an easy task since segment are stored in sorted 
+collection. Each segment may contain only single term. 
+
+For remote partitions (cloud storage) we are going to use 
+information stored in partition manifests. Manifest contains both the segment 
+term together with offsets stored in a segment. Given that we do not allow two 
```

Comment:
Is this true in compacted segments?

Reply:
yes, compacted segments end offset doesn't change, we are safe to use segment end offset as a source of offset for leader epoch.